src/track/beats.h

```diff
@@ -1,177 +1,239 @@
+                    timeSignatureMarkers =
+                            QVector<track::io::TimeSignatureMarker>());
+
+    static const QString BEAT_MAP_VERSION;
+    static const QString BEAT_GRID_1_VERSION;
+    static const QString BEAT_GRID_2_VERSION;
+    static const QString BEATS_VERSION;
+    Beat findNthBeat(FramePos frame, int offset) const;
+    Beat findNextBeat(FramePos frame) const;
+    Beat findPrevBeat(FramePos frame) const;
+    Bpm getGlobalBpm() const;
+    bool isValid() const;
+    void updateStreamInfo(const mixxx::audio::StreamInfo& streamInfo);
+    int numBeatsInRange(FramePos startFrame, FramePos endFrame) const;
+    QByteArray toProtobuf() const;
+    QString getVersion() const;
+    QString getSubVersion() const;
+    void setSubVersion(const QString& subVersion);
+    void scale(BPMScale scale);
+    FramePos findNBeatsFromFrame(FramePos fromFrame, double beats) const;
+    bool findPrevNextBeats(FramePos frame,
+            FramePos* pPrevBeatFrame,
+            FramePos* pNextBeatFrame) const;
+    void setGrid(Bpm dBpm, FramePos firstBeatFrame = kStartFramePos);
+    FramePos findClosestBeat(FramePos frame) const;
+    Bpm getBpmAtPosition(FramePos curFrame) const;
+    void setSignature(TimeSignature sig, int downbeatIndex);
+    void translate(FrameDiff_t numFrames);
+    void setBpm(Bpm bpm, int beatIndex = kFirstBeatIndex);
+    int size() const;
+    FramePos getFirstBeatPosition() const;
+    FramePos getLastBeatPosition() const;
+    Beat getBeatAtIndex(int index) const;
+    void setAsDownbeat(int beatIndex);
+
+  private:
+    void updateBpm();
+    void scaleDouble();
+    void scaleTriple();
+    void scaleQuadruple();
+    void scaleHalve();
+    void scaleThird();
+    void scaleFourth();
+    void scaleMultiple(uint multiple);
+    void scaleFraction(uint fraction);
+    void generateBeatsFromMarkers();
+    void clearMarkers();
+    SINT getSampleRate() const;
+    double getDurationSeconds() const;
+    void setFirstBeatFrame(FramePos framePos);
+    FramePos getFirstBeatFrame() const;
+
+    QString m_subVersion;
+    Bpm m_bpm;
+    BeatList m_beats;
+    track::io::Beats m_beatsProto;
+    audio::StreamInfo m_streamInfo;
+    friend QDebug operator<<(QDebug dbg, const BeatsInternal& arg);
 };
 
-// Beats is a pure abstract base class for BPM and beat management classes. It
-// provides a specification of all methods a beat-manager class must provide, as
-// well as a capability model for representing optional features.
-class Beats : public QObject {
+/// Beats is a class for BPM and beat management classes.
+/// It stores beats information including beats position, down beats position,
+/// phrase beat position and changes in tempo.
+class Beats final : public QObject {
     Q_OBJECT
+  private:
+    explicit Beats(const audio::StreamInfo& streamInfo = audio::StreamInfo(),
+            const BeatsInternal& internal = BeatsInternal());
```

Comment:
This is strange that the Beats constructor takes a BeatsInternal argument which must be constructed elsewhere. I would expect something called "BeatsInternal" to only be a private member of Beats not used anywhere outside of Beats. I am not sure if this is an issue with the code architecture or "BeatsInternal" being a misleading name. @uklotzde what do you think?

Reply:
Think of `Beats` as the socket that deals with all the outer wiring and of `BeatsInternal` as the finished bulb that is just mounted in. Assembling it in place would be too complex because then you would need to mix the wiring with the assembly parameters.

This could be a use case for move construction, i.e. *r-value reference* in conjunction with `std::forward`. Unfortunately, C++ makes this semantics tedious to write whereas in Rust it is the default for both assignment (= name binding) and passing parameters.